### PR TITLE
[Feature] Support bitmap/hll/percentile rewrite in async MV

### LINKED ISSUE #21234: Support Bitmap Rewrite in Async MV
## Enhancement

Like sync mv, we should support bitmap rewrite in async mv (rollup or not).

```
create table user_tags 
(time date, user_id int, user_name varchar(20), tag_id int) 
partition by range (time)  (partition p1 values less than MAXVALUE) distributed by hash(time) buckets 3 properties('replication_num' = '1');

create materialized view user_tags_mv1 
distributed by hash(user_id)
as select user_id, bitmap_union(to_bitmap(tag_id)) from user_tags group by user_id;

explain select user_id, bitmap_union_count(to_bitmap(tag_id)) x from user_tags group by user_id;
```

### PR DESCRIPTION
**Pull request by LiShuMing**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #21234

## Problem Summary(Required) ：
- Like Sync MV, SR can rewrite async mv: support bitmap/hll/percentile's mv rewrite.

```
String mv = "select user_id, time, bitmap_union(to_bitmap(tag_id % 10)) from user_tags group by user_id, time;";
// rewrite count distinct to bitmap_count(bitmap_union(to_bitmap(x)));
testRewriteOK(mv, "select user_id, count(distinct tag_id % 10) x from user_tags group by user_id;");
```

TODO:
- optimize rewritten plan : Convert `BITMAP_COUNT(BITMAP_UNION(TO_BITMAP)))` to `bitmap_union_count`;

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java [hunk 1]
```diff
@@ -126,7 +126,7 @@ protected OptExpression viewBasedRewrite(RewriteContext rewriteContext, OptExpre
         boolean isRollup = isRollupAggregate(mvGroupingKeys, queryGroupingKeys);
 
         // Cannot ROLLUP distinct
-        if (isRollup && !canRewriteForRollup(mvAggOp)) {
+        if (isRollup && mvAggOp.getAggregations().values().stream().anyMatch(callOp -> callOp.isDistinct())) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:129]: what if the mv's distinct columns is not used?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java [hunk 2]
```diff
@@ -152,12 +152,84 @@ protected OptExpression viewBasedRewrite(RewriteContext rewriteContext, OptExpre
                 }
                 mvOptExpr = OptExpression.create(new LogicalFilterOperator(rewrittenPred), mvOptExpr);
             }
-            return rewriteProjection(rewriteContext, queryExprToMvExprRewriter, mvOptExpr);
+            return rewriteProjection(rewriteContext, queryAggOp, queryExprToMvExprRewriter, mvOptExpr);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:155]: for bitmap/hll/percentile, it seems that we should support rollup?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+            return false;
+        }
+
+        CallOperator aggFunc = (CallOperator) op;
+        String aggFuncName = aggFunc.getFnName();
+        if (aggFuncName.equals(FunctionSet.AVG)) {
+            return true;
+        }
+        // BITMAP
+        if (aggFuncName.equals(FunctionSet.COUNT) && aggFunc.isDistinct()) {
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java:78]: use a set, put all supported funtion names into it, check with contain?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java (hunk 2 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:155]: yep. `Rollup` is supported in `rewriteAggregationOperatorByRules ` method like `avg` before. 

Here we should support `projection` also.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:129]: I just revert the codes to before... The syntax is the same with before.

I think it should not be rewritten yet even for ` the mv's distinct columns is not used`?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java (hunk 3 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java:78]: Let me think it.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java (hunk 2 above)
Comment (ABingHuang) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:155]: is Rollup with distinct supported in this pr?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java:155]: yep. only count (distinct) with rollup is supported: MV is created by bitmap_union(xxx) like single mv.

**Review by ABingHuang — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+import java.util.List;
+import java.util.Map;
+
+import static com.starrocks.catalog.Function.CompareMode.IS_IDENTICAL;
+
+/**
+ * `AggregateFunctionRewriter` will try to rewrite some agg functions to some transformations so can be
+ * better to be rewritten.
+ * eg: AVG -> SUM / COUNT
+ * eg: COUNT(DISTINCT) -> BITMAP_COUNT(BITMAP_UNION(TO_BITMAP)))
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java:42]: why not bitmap_union_count(TO_BITMAP)?

Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java:42]: `AggregatedMaterializedViewRewriter` (Multi Table Rewriter) will try rewrite the same expression between MV and query, and compensate the lost part of Query.

If MV's defined like `BITMAP_UNION(TO_BITMAP) as col1`, we try convert `COUNT(DISTINCT) -> BITMAP_COUNT(BITMAP_UNION(TO_BITMAP)))`, so later query can be rewritten by bitmap_count(col1).

File: test/sql/test_materialized_view/R/test_materialized_view_rewrite [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+2	1
+3	1
+-- !result
+select user_id, bitmap_count(bitmap_union(to_bitmap(tag_id))) x from user_tags group by user_id order by user_id;
+-- result:
+1	5
+2	1
+3	1
+-- !result
+set enable_materialized_view_rewrite = on;
```
Comment (mofeiatwork) [inline test/sql/test_materialized_view/R/test_materialized_view_rewrite:54]: this statement only turn on/off the rewrite, but without verify whether rewrite or not ?

Comment (LiShuMing) [reply test/sql/test_materialized_view/R/test_materialized_view_rewrite:54]: each agg type:
- collect result for rewrite is off;
- test no-rollup rewrite result;
- test rollup rewrite result.

NOTE: `New Tester` can not output `explain` result, just sleep 2s to ensure mv is ready.


image

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**